File: perceptron/perceptron.py
import numpy as np
import matplotlib.pyplot as plt
import random
# plt.plot([1,2,3,4])
# plt.ylabel('some numbers')
# plt.show()

MAX_ITER = 100
LEARNING_RATE = 0.1
NUM_INSTANCES = 100
theta = 0
def graph(formula, x_range):  
    x = np.array(x_range)  
    y = eval(formula)
    plt.plot(x, y)  
    

def main():
  x = []
  y = []
  truth = []

  # for ind, item in enumerate(x):
  #   x[ind] = random.uniform(0, 10)

  for i in range(50):
    x.append(random.uniform(0, 10))
    y.append(random.uniform(10, 20))
    truth.append(1)

  for j in range(50, 100):
    x.append(random.uniform(10, 20))
    y.append(random.uniform(0, 10))
    truth.append(0)

  weights = [random.uniform(0, 1), random.uniform(0, 1), random.uniform(0, 1)]

  iteration = 0
  # print(cal_output(0, [1,2,3], 1, 2))
  local_error = 0
  global_error = 0
  output = 0
  should_stop = False
  while(True):
   
    iteration += 1
    global_error = 0

    for ind in range(NUM_INSTANCES):
      output = cal_output(theta, weights, x[ind], y[ind])
      local_error = truth[ind] - output
      # print(local_error)
      weights[0] += LEARNING_RATE * local_error * x[ind]
      weights[1] += LEARNING_RATE * local_error * y[ind]
      weights[2] += LEARNING_RATE * local_error
      global_error += (local_error * local_error)
    print("RMS Error: %n", global_error)
    if (global_error == 0 or iteration > MAX_ITER):

      break

      
  str = "{}*x + {}*y + {} = 0"
  print(str.format(weights[0], weights[1], weights[2]))
  equation = "-({}/{})*x-{}/{}"
  final_equation = equation.format(weights[0], weights[1], weights[2], weights[1])
  print(final_equation)
  graph(final_equation, range(0,20))

    # if (test == 10):
    #   break
    # test += 1
    # print(test)


  



  plt.plot(x, y, 'ro')
  plt.axis([0, 20, 0, 20])

  plt.show()

def cal_output(t, w, x, y):
  sum = x * w[0] + y * w[1] + w[2]
  if (sum >= t):
    return 1
  else:
    return 0


main()

File: perceptron/perceptron.java
import java.util.*;
import java.io.*;
import java.text.*;
import java.math.*;

class Perception {

	static int MAX_ITER = 100;
	static int NUM_INSTANCES = 100;
	static int theta = 0;
	static double LEARNING_RATE = 0.1;

	public static void main(String args[]) {
		double[] x = new double [NUM_INSTANCES];
		double[] y = new double [NUM_INSTANCES];

		int[] outputs = new int [NUM_INSTANCES];

		for (int i = 0; i < NUM_INSTANCES / 2; i++) {
			x[i] = genRandom(0, 10);
			y[i] = genRandom(10, 20);

			outputs[i] = 0;
		}

		for (int i = 50; i < NUM_INSTANCES; i++) {
			x[i] = genRandom(10, 20);
			y[i] = genRandom(0, 10);

			outputs[i] = 1;
		}

		double[] weights = new double[4];
		double localError, globalError;
		int i, p, iteration, output;

		weights[0] = genRandom(0, 1);
	  weights[1] = genRandom(0, 1);	
	  weights[2] = genRandom(0, 1); //bias

	  iteration = 0;
	  do {
	  	iteration ++;
	  	globalError = 0;

	  	for (p = 0; p < NUM_INSTANCES; p++) {
	  		output = calOutput(theta, weights, x[p], y[p]);
	  		localError = outputs[p] - output;

	  		weights[0] += LEARNING_RATE * localError * x[p];
	  		weights[1] += LEARNING_RATE * localError * y[p];
	  		weights[2] += LEARNING_RATE * localError;
	  		globalError += (localError=localError);
	  	}
	  	System.out.println("Iteration "+iteration+" : RMSE = "+Math.sqrt(globalError/NUM_INSTANCES));

	  } while (globalError != 0 && iteration<=MAX_ITER);

	  System.out.println("\n=======\nDecision boundary equation:");
		System.out.println(weights[0] +"*x + "+weights[1]+"*y +  " + weights[3]+" = 0");


	}

	public static double genRandom(int min, int max) {
		DecimalFormat df = new DecimalFormat("#.####");
		double d = min + Math.random() * (max - min);
		String s = df.format(d);
		double x = Double.parseDouble(s);
		return x;
	}
	static int calOutput(int theta, double weights[], double x, double y) {
		double sum = x * weights[0] + y * weights[1] + weights[2];
		return (sum >= theta) ? 1 : 0;
	}
}